File: ExtraiPapelSeriesHistoricasB3.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;

public class ExtraiPapelSeriesHistoricasB3 {
    
    public static void main(String[] args) {
        System.out.println("Uso: java ExtraiPapelSeriesHistoricasB3 <ARQUIVO-COTAHIST_AAAA.TXT> <PAPEIS> [NOMEARQUIVOSAIDA]");
        System.out.println("Extrai de um arquivo no formato B3 de séries históricas de cotações, uma lista com alguns papéis.");
        System.out.println("<PAPEIS> pode ser um ticker específico, ou pode ser uma lista de papéis separados por ; Ex: BBAS3;KNRI11;BBDC4");
        System.out.println("Gera um arquivo texto separado por ; com o nome do papel, ou com o nome passado no parâmetro NOMEARQUIVOSAIDA.");
        
        String arqB3 = null, arqOut = null, argPapel = null, papeis[] = null;
        try {
            arqB3 = args[0];
            argPapel = args[1].trim();
            if (args.length > 2) {
                arqOut = args[2];
            }
            
            String[] aux = argPapel.split(";");
            int count = 0;
            for (String pap: aux) {
                if (pap != null && !pap.equals("")) {
                    count++;
                }
            }
            papeis = new String[count];
            count = 0;
            for (String pap: aux) {
                if (pap != null && !pap.equals("")) {
                    papeis[count] = pap;
                    count++;
                }
            }
        }
        catch (Exception e) {
            System.out.println("Argumentos inválidos.");
            return;
        }
        
        FileWriter writer = null;
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(arqB3)))) {            
            String cabecalho = reader.readLine();
            if (cabecalho.startsWith("00COTAHIST")) {
                String ano = cabecalho.substring(11, 15);
                if (arqOut == null) {
                    arqOut = argPapel.replace(";", "") + ano + ".csv";
                }
                    
                writer = new FileWriter(arqOut);
                //cabecalho de saida
                writer.write("Papel;Data;PrecoAbertura;PrecoFechamento;PrecoMaximo;PrecoMinimo;PrecoMedio;PrecoMelhorOfertaCompra;PrecoMelhorOfertaVenda;NrNegocios;QtdeTotalPapeisNegociados;VolumeNegociado;\n");
                    
                    String linha = reader.readLine();
                    while (linha != null && !linha.startsWith("99COTAHIST.")) {
                        String codigo = linha.substring(12, 24).trim();
                        boolean selecionado = false;
                        for (String p: papeis) {
                            if (p.equalsIgnoreCase(codigo)) {
                                selecionado = true;
                                break;
                            }
                        }
                        if (selecionado) {
                            String data = linha.substring(2, 10);
                            data =  data.substring(6, 8) + "/" + data.substring(4, 6) + "/" + data.substring(0, 4); 
                        
                        String abertura = linha.substring(56, 67) + "," + linha.substring(67, 69);
                        String maximo = linha.substring(69, 80) + "," + linha.substring(80, 82);
                        String minimo = linha.substring(82, 93) + "," + linha.substring(93, 95);
                        String medio = linha.substring(95, 106) + "," + linha.substring(106, 108);
                        String fechamento = linha.substring(108, 119) + "," + linha.substring(119, 121);
                        String melhorOfertaCompra = linha.substring(121, 132) + "," + linha.substring(132, 134);
                        String melhorOfertaVenda = linha.substring(134, 145) + "," + linha.substring(145, 147);
                        String negocios = linha.substring(147, 152);
                        String papeisNegociados = linha.substring(152, 170);
                        String volume = linha.substring(170, 186) + "," + linha.substring(186, 188);
                        
                        String toWrite = codigo + ";" + data + ";" + abertura + ";" + fechamento + ";" + maximo + ";" + minimo + ";" + medio + ";" + 
                        melhorOfertaCompra + ";" + melhorOfertaVenda + ";" + negocios + ";" + papeisNegociados + ";" + volume + ";\n";
                        writer.write(toWrite);
                    }                                
                    linha = reader.readLine();
                }
                
                reader.close();
                writer.close();
            }
            else {
                System.out.println("Formato arquivo inválido.");
                return;
            }            
        }
        catch (Exception e) {
            try { if (writer != null) writer.close(); } catch (IOException e1) { e1.printStackTrace(); }
            e.printStackTrace();
        }
        
        System.out.println("Terminado.");
    }

}

File: ExtraiPapelSeriesHistoricasB3.py
import os.path
from datetime import datetime

def extrair_cotacoes(nome_arquivo, ativos, nome_arquivo_saida):
    file1 = open(nome_arquivo, 'r')
    file2 = open(nome_arquivo_saida, 'a')
    
    cabecalho = file1.readline()
    if cabecalho.startswith('00COTAHIST'): 
        while True:
            linha = file1.readline()
            if not linha:
                break

            #print(linha)

            #fim de arquivo
            if linha.startswith('99COTAHIST'): 
                break;

            codigo = linha[12:24].strip();
            if codigo in ativos:
                linha_a_escrever = parse_linha(linha)
                file2.write(linha_a_escrever)
                #print('WRITING: ' + linha_a_escrever)
    else:
        print('Formato de arquivo inválido.')

    file1.close()
    file2.close()


def extrair_cotacoes_arquivo(nome_arquivo, ativos, nome_arquivo_saida):
    #cria arquivo novo de saida e escreve cabecalho
    file2 = open(nome_arquivo_saida, 'w')
    file2.write('Papel;Data;PrecoAbertura;PrecoFechamento;PrecoMaximo;PrecoMinimo;PrecoMedio;PrecoMelhorOfertaCompra;PrecoMelhorOfertaVenda;NrNegocios;QtdeTotalPapeisNegociados;VolumeNegociado;\n');
    file2.close()
    extrair_cotacoes(nome_arquivo, ativos, nome_arquivo_saida)
    print('Finished')


def extrair_cotacoes_varios_arquivos(nome_base, ano_inicial, ativos, nome_arquivo_saida):
    #cria arquivo novo de saida e escreve cabecalho
    file2 = open(nome_arquivo_saida, 'w')
    file2.write('Papel;Data;PrecoAbertura;PrecoFechamento;PrecoMaximo;PrecoMinimo;PrecoMedio;PrecoMelhorOfertaCompra;PrecoMelhorOfertaVenda;NrNegocios;QtdeTotalPapeisNegociados;VolumeNegociado;\n');
    file2.close()

    ano = ano_inicial
    ano_atual = datetime.now().year
    while ano <= ano_atual:
        print('Processando ano ' + str(ano))
        nome_arquivo = nome_base + str(ano) + '.txt'
        extrair_cotacoes(nome_arquivo, ativos, nome_arquivo_saida)
        ano = ano + 1

    print('Finished')


def parse_linha(linha):
    data = linha[2:10]
    data =  data[6:8] + '/' + data[4:6] + '/' + data[0:4] 
    codigo = linha[12:24].strip();
    
    abertura = linha[56:67] + ',' + linha[67:69]
    maximo = linha[69:80] + ',' + linha[80:82]
    minimo = linha[82:93] + ',' + linha[93:95]
    medio = linha[95:106] + ',' + linha[106:108]
    fechamento = linha[108:119] + ',' + linha[119:121]
    melhor_oferta_compra = linha[121:132] + ',' + linha[132:134]
    melhor_oferta_venda = linha[134:145] + ',' + linha[145:147]
    negocios = linha[147:152]
    papeis_negociados = linha[152:170]
    volume = linha[170:186] + ',' + linha[186:188]
    
    to_write = codigo + ';' + data + ';' + abertura + ';' + fechamento + ';' + maximo + ';' + minimo + ';' + medio + ';' 
    to_write = to_write + melhor_oferta_compra + ';' + melhor_oferta_venda + ';' + negocios + ';' + papeis_negociados + ';' + volume + ';\n';
    return to_write;


#Exemplos de chamadas:
#extrair_cotacoes_arquivo('D:\\COTAHIST_A2012.TXT', ['ABCP11', 'KNRI11'], 'D:\\resultados3.csv')
extrair_cotacoes_varios_arquivos('D:\\COTAHIST_A', 2019, 'KNRI11', 'D:\\resultados4.csv')
#
#Para executar, basta modificar alguma dessas chamadas de função em uma nova linha sem # 
#Salvar o arquivo e executar o comando py ExtraiPapelSeriesHistoricasB3.py
#
